### DISCUSSION TITLE
cannot set breakpoint at _el2_entry when using gdb debug of bao

### DISCUSSION THREAD

**Discussion by xioliu** [bao-project/bao-hypervisor#143] [General]
Hello,

I am running the baremetal demo of bao, when I tried to set the breakpoint to _el2_entry or 0xfd8000000000, gdb cannot stop at the place. I found the earliest breakpoint I can set successfully is init in init.c.
Could you please give me any clue that I can debug the boot code from _el2_entry?
Thank you.

**Comment by DavidMCerdeira** [comment]
Hi,

You are not able to set the breakpoint at that point because those virtual memory memory addresses are only used starting from the `init` in init.c

The way to debug that code is to put a breakpoint on the physical address bao is running on.
If you are following [these steps](https://github.com/bao-project/bao-demos/blob/master/platforms/qemu-aarch64-virt/README.md) you should put a breakpoint at address 0x50000000

Let me know if this helps.

**Comment by xioliu** [comment]
Hello David,

Thank you. 
I believe 0x50000000 is the address of _start in bare metal image, and yes, I can set breakpoint at 0x50000000 if I use "aarch64-none-elf-gdb baremetal.elf --tui" to debug. However, I want to debug the bao.elf and start from _el2_entry.

**Reply by DavidMCerdeira** [reply]
That should not be an issue for your purposes. Setting the breakpoint at 0x50000000 may cause the execution to break in the baremetal app but it will also break the execution at _el2_entry.

I'm assuming you are trying to use the _el2_entry symbol to create the breakpoint. This will not work, as the generated elf will set this symbol address as part of the virtual address space, which is only set just before the init.
You should set the breakpoint in gdb with `break *0x50000000` or  `break *(unsigned long*)0x50000000` depending on gdb version.
This will break execution and let you see the disassembly. 

If you want to sync with assembly source code you will have to tell gdb the address where you want to load the symbols to.
Let me know if you need help doing this.

**Reply by xioliu** [reply]
I tried below commands:
export PATH=$PATH:/home/xioliu/arm-gnu-toolchain-13.2.Rel1-x86_64-aarch64-none-elf/bin
aarch64-none-elf-gdb bao.elf --tui

In gdb:
b *0x50000000
add-symbol-file bao.elf 0x50000000
add-symbol-file baremetal.elf 0x50000000
info b # showed the breakpoint in exceptions.S:71, which means vcpu_arch_entry. 

I don't understand why the break point is set here. Though after running, it did stop here.
Another thing is, the address of registers are not as defined in bao.elf.txt.
![image](https://github.com/bao-project/bao-hypervisor/assets/95175997/d02321d0-31bd-4904-8055-097600152001)

Could you explain more? Thank you.

**Reply by DavidMCerdeira** [reply]
Greetings,

Sure.
I'll try to reproduce your case by the end of the week to get you accurate feedback.

**Reply by DavidMCerdeira** [reply]
Hi,
Sorry for the delay.
We should provide an offset for all symbols. 
The offset is: bao base address - load address.
You can find the base address by running
```
readelf -We bao/bin/qemu-aarch64-virt/baremetal/bao.elf | grep .boot
```
In my case this outputs:
```
$ readelf -We bao/bin/qemu-aarch64-virt/baremetal/bao.elf | grep .boot
  [ 1] .boot             PROGBITS        0000fd8000000000 001000 000540 00  AX  0   0  8
   00     .boot .text .rodata
```

So our calculation is as follows:
0xFD8000000000−0x50000000 = 0xFD7FB0000000

Now, if you run the following command, you should be able to debug the init code:
```
add-symbol-file bin/qemu-aarch64-virt/baremetal/bao.elf -o -FD7FB0000000
```

**Reply by xioliu** [reply]
Thank you David.

I tried "add-symbol-file bao.elf -o 0xFD7FB0000000", however I could not see the symbols and could not debug.
![image](https://github.com/bao-project/bao-hypervisor/assets/95175997/e280e562-b2ae-4ee3-b952-8c610850d9af)

**Reply by xioliu** [reply]
Normally I use "b init" just after "target remote:1234" and all things are good except not being able to trace the code before init.
![image](https://github.com/bao-project/bao-hypervisor/assets/95175997/4d5d7a15-39d2-4ec7-a072-6982988fb843)

**Reply by xioliu** [reply]
I am wondering the reason might be, init is the earliest function gdb can find in the symbol, other functions such as boot_cache_invalidate and _el2_entry/_reset_handler are defined as NOTYPE in bao.elf.txt.

**Reply by xioliu** [reply]
> I am wondering the reason might be, init is the earliest function gdb can find in the symbol, other functions such as boot_cache_invalidate and _el2_entry/_reset_handler are defined as NOTYPE in bao.elf.txt.

Before, _el2_entry is NOTYPE in bao.elf.txt
1307: 0000fd8000000000     0 NOTYPE  GLOBAL DEFAULT    1 _el2_entry

After adding ".type _el2_entry, @function" in boot.S  it changes to FUNC type, and gdb can set breakpoint on it. However there is still no source available in gdb window and dgb does not stop at _el2_entry.
  1307: 0000fd8000000000     0 FUNC    GLOBAL DEFAULT    1 _el2_entry
![image](https://github.com/bao-project/bao-hypervisor/assets/95175997/76f90dec-58a2-4653-afec-ac8d14f48191)

**Reply by DavidMCerdeira** [reply]
It seems we are observing different gdb behavior.

When I add the symbols with the offset right after starting gdb:
```
add-symbol-file bin/qemu-aarch64-virt/baremetal/bao.elf -o -FD7FB0000000
add symbol table from file "bin/qemu-aarch64-virt/baremetal/bao.elf" with all sections offset by 0xffff028050000000
```

I noticed the my offset value is different than yours.

When you do:
```
export PATH=$PATH:/home/xioliu/arm-gnu-toolchain-13.2.Rel1-x86_64-aarch64-none-elf/bin
```
I think you may be using the wrong gdb version.
IRC how the PATH works in the shell, you're putting that toolchain in the last place the shell checks to find a binary named `aarch64-none-elf-gdb`
You can try to use the absolute path or set the path as follows:
```
export PATH=/home/xioliu/arm-gnu-toolchain-13.2.Rel1-x86_64-aarch64-none-elf/bin:$PATH
```

**Reply by xioliu** [reply]
Sorry for late feedback, it has been hectic weeks.
The only aarch64-none-elf-gdb on ubuntu is under /home/xioliu/arm-gnu-toolchain-13.2.Rel1-x86_64-aarch64-none-elf/bin, so I think the PATH export is not the cause.

**Reply by DavidMCerdeira** [reply]
I just realized you are missing a minus sign when you do this: 
```
add-symbol-file bao.elf -o 0xFD7FB0000000 
```
Should be:
```
add-symbol-file bao.elf -o -0xFD7FB0000000 
```
Let me know if this works

**Reply by xioliu** [reply]
@DavidMCerdeira Thank you so much David. It works.
